src/components/views/elements/AppTile.js

```diff
@@ -98,23 +103,35 @@ export default React.createClass({
         return false;
     },
 
-    componentWillMount: function() {
+    componentWillMount() {
+        window.addEventListener('message', this._onMessage, false);
+        this.updateWidgetContent();
+    },
+
+    // Update widget content
+    updateWidgetContent() {
+        this.setState(this._getInitialState());
+        // Set scalar token on the wUrl, if needed
+        this.setScalarToken();
+    },
+
+    // Adds a scalar token to the widget URL, if required
+    setScalarToken() {
         if (!this.isScalarUrl()) {
             return;
         }
         // Fetch the token before loading the iframe as we need to mangle the URL
-        this.setState({
-            loading: true,
-        });
-        this._scalarClient = new ScalarAuthClient();
+        if(! this._scalarClient) {
+            this._scalarClient = new ScalarAuthClient();
+        }
         this._scalarClient.getScalarToken().done((token) => {
-            // Append scalar_token as a query param
+            // Append scalar_token as a query param if not already present
             this._scalarClient.scalarToken = token;
             const u = url.parse(this.props.url);
-            if (!u.search) {
-                u.search = "?scalar_token=" + encodeURIComponent(token);
-            } else {
-                u.search += "&scalar_token=" + encodeURIComponent(token);
+            const params = new URLSearchParams(u.search);
```

Comment:
Looks like we already have a library for this - riot-web uses `querystring` (looking at `url_utils.js`)... although it doesn't seem to be marked as a dependency.

Reply:
Urghhh... undocumented dependencies. Yay. I'll switch to using that then. I'll also add querystring in to the package.json to make it more obvious that this library is available.